<a href="https://colab.research.google.com/github/teamnassim/AAI510/blob/master/Convolutional_Neural_Networks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Import the keras libraries and the packages that we would need to build a neural network.


!pip install numpy==1.21.4
!pip install pandas==1.3.4
!pip install keras==2.1.6
!pip install tensorflow==1.15

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 15.7 MB 4.4 MB/s 
  Attempting uninstall: numpy
    Found existing installation: numpy 1.21.6
    Uninstalling numpy-1.21.6:
      Successfully uninstalled numpy-1.21.6
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.
albumentations 0.1.12 requires imgaug<0.2.7,>=0.2.5, but you have imgaug 0.2.9 which is incompatible.


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 11.3 MB 4.2 MB/s 
  Attempting uninstall: pandas
    Found existing installation: pandas 1.3.5
    Uninstalling pandas-1.3.5:
      Successfully uninstalled pandas-1.3.5


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 339 kB 4.9 MB/s 
  Attempting uninstall: keras
    Found existing installation: keras 2.8.0
    Uninstalling keras-2.8.0:
      Successfully uninstalled keras-2.8.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.8.2+zzzcolab20220527125636 requires keras<2.9,>=2.8.0rc0, but you have keras 2.1.6 which is incompatible.


In [1]:
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

Using TensorFlow backend.


When working with convolutional neural networks in particular, we will need additional packages.


In [2]:
from keras.layers.convolutional import Conv2D # to add convolutional layers
from keras.layers.convolutional import MaxPooling2D # to add pooling layers
from keras.layers import Flatten # to flatten data for fully connected layers

<a id='item42'></a>


## Convolutional Layer with One set of convolutional and pooling layers


In [3]:
# import data
from keras.datasets import mnist

# load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# reshape to be [samples][pixels][width][height]
X_train = X_train.reshape(X_train.shape[0], 28, 28, 1).astype('float32')
X_test = X_test.reshape(X_test.shape[0], 28, 28, 1).astype('float32')

## Normalize the pixel values to be between 0 and 1


In [4]:
X_train = X_train / 255 # normalize training data
X_test = X_test / 255 # normalize test data

convert the target variable into binary categories


In [5]:
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

num_classes = y_test.shape[1] # number of categories

Define a function that creates our model. Let's start with one set of convolutional and pooling layers.


In [6]:
def convolutional_model():
    
    # create model
    model = Sequential()
    model.add(Conv2D(16, (5, 5), strides=(1, 1), activation='relu', input_shape=(28, 28, 1)))
    model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
    
    model.add(Flatten())
    model.add(Dense(100, activation='relu'))
    model.add(Dense(num_classes, activation='softmax'))
    
    # compile model
    model.compile(optimizer='adam', loss='categorical_crossentropy',  metrics=['accuracy'])
    return model

Call the function to create the model, and then let's train it and evaluate it.


In [7]:
# build the model
model = convolutional_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=200, verbose=2)

# evaluate the model
scores = model.evaluate(X_test, y_test, verbose=0)
print("Accuracy: {} \n Error: {}".format(scores[1], 100-scores[1]*100))







Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Train on 60000 samples, validate on 10000 samples
Epoch 1/10





 - 27s - loss: 0.2795 - acc: 0.9242 - val_loss: 0.0888 - val_acc: 0.9735
Epoch 2/10
 - 20s - loss: 0.0780 - acc: 0.9772 - val_loss: 0.0598 - val_acc: 0.9817
Epoch 3/10
 - 21s - loss: 0.0546 - acc: 0.9838 - val_loss: 0.0470 - val_acc: 0.9856
Epoch 4/10
 - 20s - loss: 0.0418 - acc: 0.9879 - val_loss: 0.0399 - val_acc: 0.9869
Epoch 5/10
 - 21s - loss: 0.0340 - acc: 0.9893 - val_loss: 0.0367 - val_acc: 0.9882
Epoch 6/10
 - 21s - loss: 0.0282 - acc: 0.9915 - val_loss: 0.0376 - val_acc: 0.9874
Epoch 7/10
 - 20s - loss: 0.0228 - acc: 0.9929 - val_loss: 0.0411 - val_acc: 0.9862
Epoch 8/10
 - 20s - loss: 0.0201 - acc: 0.9936 - val_loss: 0.0358 - val_acc: 0.9893
Epoch 9/10
 - 20s - loss: 0.0158 - acc: 0.9950 - val_loss: 0.0337 - val_acc: 0.9887
Epoch 10/10
 - 23s - loss: 0.0132 - acc: 0.9963 - val_loss: 0.0352 - val_acc: 0.9887
A

***


<a id='item43'></a>


## Convolutional Layer with two sets of convolutional and pooling layers


Redine the convolutional model so that it has two convolutional and pooling layers instead of just one layer of each.


In [8]:
def convolutional_model():
    
    # create model
    model = Sequential()
    model.add(Conv2D(16, (5, 5), activation='relu', input_shape=(28, 28, 1)))
    model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
    
    model.add(Conv2D(8, (2, 2), activation='relu'))
    model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
    
    model.add(Flatten())
    model.add(Dense(100, activation='relu'))
    model.add(Dense(num_classes, activation='softmax'))
    
    # Compile model
    model.compile(optimizer='adam', loss='categorical_crossentropy',  metrics=['accuracy'])
    return model

Now, let's call the function to create our new convolutional neural network, and then let's train it and evaluate it.


In [9]:
# build the model
model = convolutional_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=200, verbose=2)

# evaluate the model
scores = model.evaluate(X_test, y_test, verbose=0)
print("Accuracy: {} \n Error: {}".format(scores[1], 100-scores[1]*100))

Train on 60000 samples, validate on 10000 samples
Epoch 1/10
 - 24s - loss: 0.5027 - acc: 0.8555 - val_loss: 0.1425 - val_acc: 0.9551
Epoch 2/10
 - 22s - loss: 0.1217 - acc: 0.9643 - val_loss: 0.0809 - val_acc: 0.9735
Epoch 3/10
 - 23s - loss: 0.0863 - acc: 0.9747 - val_loss: 0.0657 - val_acc: 0.9789
Epoch 4/10
 - 22s - loss: 0.0687 - acc: 0.9794 - val_loss: 0.0584 - val_acc: 0.9816
Epoch 5/10
 - 22s - loss: 0.0590 - acc: 0.9825 - val_loss: 0.0473 - val_acc: 0.9840
Epoch 6/10
 - 22s - loss: 0.0520 - acc: 0.9846 - val_loss: 0.0523 - val_acc: 0.9846
Epoch 7/10
 - 22s - loss: 0.0469 - acc: 0.9857 - val_loss: 0.0391 - val_acc: 0.9864
Epoch 8/10
 - 23s - loss: 0.0417 - acc: 0.9873 - val_loss: 0.0396 - val_acc: 0.9863
Epoch 9/10
 - 22s - loss: 0.0391 - acc: 0.9881 - val_loss: 0.0449 - val_acc: 0.9845
Epoch 10/10
 - 22s - loss: 0.0357 - acc: 0.9894 - val_loss: 0.0393 - val_acc: 0.9870
Accuracy: 0.987 
 Error: 1.2999999999999972


# Summary

I used two models to test the MNIST dataset: Convolutional layer with 1 set of convolutional and pooling layers & Convolutional layer with 2 sets of convolutional and pooling layers. I began with both loading the data and preparing it. First, I reshaped the parameters to be [samples][pixels][width][height]. I then normalized the pixel values to be between 0 and 1. Lastly, I converted the target variable into binary categories. After preparing the data, I created and built the model with the layers. I used the Convolutional layer (with ReLu activation), pooling layer, and the flattening layer with ReLu and softmax activation. For compiling the model, I compiled the network by providing 3 parameters: the optimizer, loss function, and the accuracy metric that we will monitor. Adam is used as the optimization algorithm as it is efficient and best suited for problems with a lot of parameters. For the loss function, I have used the Cross-entropy loss function as we are dealing with a multiclass classification problem. This function effectively computes the difference between the actual output and the output predicted by the model which in turn is the error in the prediction. This error is given as feedback to the Adam optimizer which comes up with a better model in every iteration. For the metric to be monitored, I have used the accuracy metric. Lastly, I then fit and trained the model with 10 epochs. For the results of the 1st model, the accuracy was 0.9887 and the Error was 1.299955 for the model with 1 set of convolutional and pooling layers. For the results for the model of 2 sets of convolutional and pooling layers, the outcome was very similar to the 1st model with an accuracy score of 0.987 and an Error of 1.2999972